src/main/java/com/microsoft/sqlserver/jdbc/ConfigurableRetryLogic.java

```diff
@@ -0,0 +1,206 @@
+                }
+            } else {
+                stmtRules.put(Integer.parseInt(rule.getError()), rule);
+            }
+        }
+    }
+
+    private static String getCurrentClassPath() throws SQLServerException {
+        String location = "";
+        String className = "";
+
+        try {
+            className = new Object() {}.getClass().getEnclosingClass().getName();
+            location = Class.forName(className).getProtectionDomain().getCodeSource().getLocation().getPath();
+            location = location.substring(0, location.length() - 16);
+            URI uri = new URI(location + FORWARD_SLASH);
+            return uri.getPath() + DEFAULT_PROPS_FILE; // For now, we only allow "mssql-jdbc.properties" as file name.
+        } catch (URISyntaxException e) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_AKVURLInvalid"));
+            Object[] msgArgs = {location + FORWARD_SLASH};
+            throw new SQLServerException(form.format(msgArgs), null, 0, e);
+        } catch (ClassNotFoundException e) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_UnableToFindClass"));
+            Object[] msgArgs = {className};
+            throw new SQLServerException(form.format(msgArgs), null, 0, e);
+        }
+    }
+
+    private static LinkedList<String> readFromFile() throws SQLServerException {
+        String filePath = getCurrentClassPath();
+        LinkedList<String> list = new LinkedList<>();
+
+        try {
+            File f = new File(filePath);
+            try (BufferedReader buffer = new BufferedReader(new FileReader(f))) {
+                String readLine;
+                while ((readLine = buffer.readLine()) != null) {
+                    if (readLine.startsWith(RETRY_EXEC)) {
+                        String value = readLine.split(EQUALS_SIGN)[1];
+                        Collections.addAll(list, value.split(SEMI_COLON));
+                    }
+                }
+            }
+            timeLastModified = f.lastModified();
+        } catch (FileNotFoundException e) {
+            // If the file is not found either A) We're not using CRL OR B) the path is wrong. Do not error out, instead
+            // log a message.
+            if (CONFIGURABLE_RETRY_LOGGER.isLoggable(java.util.logging.Level.FINER)) {
+                CONFIGURABLE_RETRY_LOGGER.finest("No properties file exists or the file path is incorrect.");
+            }
+        } catch (IOException e) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_errorReadingStream"));
+            Object[] msgArgs = {e.toString()};
+            throw new SQLServerException(form.format(msgArgs), null, 0, e);
+        }
+        return list;
+    }
+
+    ConfigurableRetryRule searchRuleSet(int ruleToSearchFor) throws SQLServerException {
+        refreshRuleSet();

```

Comment:
I don't see any issues with `refreshRuleSet` and threading mainly because of the 30 sec requirement. We read not only when the file is updated, but if 30 secs has passed. So therefore, a user should not expect changes in the props file to be immediately reflected. So if a thread reads one set of props, and another thread another, i) the "wrong" thread will be updated in the next run through, and ii) this means one extra statement retry, which in comparison to the 30 sec wait is inconsequential. That is, for that thread, the change in props was reflected 30 secs and some milliseconds after file change, as opposed to 30 secs.

For the members, I see nothing that would cause issues. As most, an out of sync `timeLastModified` would lead to an extra read, which is fine.